### PR TITLE
#9493 Add Image Button to Rich Text Editor

### LINKED ISSUE #9493: Slice 2: Add image button to Rich Text Editor
**Implementation Sketch**

https://www.notion.so/pixiebrix/Phase-2-Implement-Image-Upload-in-Rich-Text-Editor-12643b21a25380eb9508e2f8419317ba?pvs=4#13b43b21a253804585aeef1e92461392

### PR DESCRIPTION
**Pull request by mnholtz**
## What does this PR do?

- Closes #9493

## Discussion

- Putting bulleted list buttons underneath the overflow menu in favor of showing the image buttton

## Demo

https://www.loom.com/share/e0761bba8c0244dd98223f02a346f86b

### CODE REVIEW

File: applications/browser-extension/src/data/service/api.ts [hunk 1]
```diff
@@ -499,6 +502,49 @@ export const appApi = createApi({
...
+    }),
+    updateAsset: builder.mutation<Asset, { databaseId: UUID; assetId: UUID }>({
+      query({ databaseId, assetId }) {
+        return {
+          url: API_PATHS.ASSET(databaseId, assetId),
+          method: "patch",
+        };
+      },
+      invalidatesTags(result, error, { assetId }) {
+        if (isAxiosError(error) && error.response?.status === 400) {
```
Comment (twschiller) [inline applications/browser-extension/src/data/service/api.ts:526]: Why is this extra invalidation logic necessary?

Comment (mnholtz) [reply applications/browser-extension/src/data/service/api.ts:526]: This was a pattern that I saw in api.ts with other endpoints; I mulling over whether or not it made sense to include for this endpoint and forgot to dig into whether or not I agreed. It seems that the reasoning is that you wouldn't want to invalidate tags if a mutation was not successful

Comment (twschiller) [reply applications/browser-extension/src/data/service/api.ts:526]: I'd be surprised if RTK Query invalidatesTags runs on 4XX errors. In any case, it should be safe to invalidate even if there's a 4XX/5XX. (The extension code doesn't use the list endpoint)

File: package-lock.json
Comment (grahamlangford) [inline package-lock.json:1]: I think something went wrong with your package-lock

File: applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { useCreateAssetPreUploadMutation } from "@/data/service/api";
+import { type UUID } from "@/types/stringTypes";
+
+const useUploadAsset: () => (
```
[the remark this answers is not in the document]
Comment (mnholtz) [reply applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts:21]: Is your concern early abstraction? I wasn't sure how complicated it was going to get, and decided to extract it early. I'm assuming your suggestion is to put it in the ImageButton component file?

[the remark this answers is not in the document]
Comment (grahamlangford) [reply applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts:21]: Yes. Avoid Hasty Abstractions

[the remark this answers is not in the document]
Comment (mnholtz) [reply applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts:82]: @grahamlangford lmk what your feedback/suggestion is now that the hook logic has been implemented. I don't know that I prefer putting this hook in the same file as ImageButton.

File: applications/browser-extension/src/types/swagger.ts
Comment (mnholtz) [inline applications/browser-extension/src/types/swagger.ts:1]: The swagger file hasn't been updated in a while

File: applications/browser-extension/src/data/model/Asset.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  };
+}
+
+export function transformAssetPreUploadResponse(
+  response: components["schemas"]["AssetPreUpload"],
+): AssetPreUpload {
+  return {
+    asset: transformAssetResponse(response.asset),
+    uploadUrl: new URL(response.upload_url),
+    // @ts-expect-error -- the type of fields in swagger is wrong (TODO make a ticket for this)
```
[the remark this answers is not in the document]
Comment (johnnymetz) [reply applications/browser-extension/src/data/model/Asset.ts:61]: Ok got it. Looks like the `fields` param is wrong.

[the remark this answers is not in the document]
Comment (johnnymetz) [reply applications/browser-extension/src/data/model/Asset.ts:61]: Fix in [this commit](https://github.com/pixiebrix/pixiebrix-app/pull/5970/commits/9997e4cc6d843978a588bd6ab9337692f2d80420). Looks like drf-spectacular doesn't have this bug (see [here](https://github.com/oxan/djangorestframework-dataclasses?tab=readme-ov-file#schemas)) so another +1 for this ticket: https://github.com/pixiebrix/pixiebrix-app/issues/5773

File: applications/browser-extension/src/data/model/Asset.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  uploadUrl: URL;
+  fields: Record<string, string>;
+};
+
+export function transformAssetResponse(
+  response: components["schemas"]["Asset"],
+): Asset {
+  return {
+    id: validateUUID(response.id),
+    // @ts-expect-error -- TODO: these don't seem to be actually optional, how have we dealt with this in the past?
```
Comment (twschiller) [inline applications/browser-extension/src/data/model/Asset.ts:42]: Instead of suppressing errors on all fields, you could use `Required<>` or `SetRequired<>` on the parameter and cast at the call-site with a comment pointing to a TODO

Comment (mnholtz) [reply applications/browser-extension/src/data/model/Asset.ts:42]: Suppressed temporarily here to decide the best course of action, will definitely be handling this without suppressing with a link to a ticket

File: applications/browser-extension/src/components/richTextEditor/toolbar/ToolbarOverflow.tsx [hunk 5]
```diff
@@ -24,11 +24,15 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
 import styles from "@/components/richTextEditor/toolbar/Toolbar.module.scss";
+import BulletedListButton from "@/components/richTextEditor/toolbar/BulletedListButton";
+import NumberedListButton from "@/components/richTextEditor/toolbar/NumberedListButton";
 
 const OverflowPopover = () => (
   <ButtonGroup size="sm">
     <UnderlineButton />
     <StrikethroughButton />
+    <BulletedListButton />
+    <NumberedListButton />
```
Comment (mnholtz) [inline applications/browser-extension/src/components/richTextEditor/toolbar/ToolbarOverflow.tsx:35]: The image button made the toolbar too wide without wrapping; we can always revisit making the overflow behavior more dynamic

File: applications/browser-extension/src/components/richTextEditor/toolbar/Toolbar.module.scss [hunk 6]
```diff
@@ -44,11 +44,7 @@
 .toolbar {
   border-bottom: 1px solid #ced4da;
   display: flex;
-  justify-content: space-between;
-
-  @media (min-width: 400px) {
-    justify-content: flex-start;
-  }
+  justify-content: start;
```
Comment (mnholtz) [inline applications/browser-extension/src/components/richTextEditor/toolbar/Toolbar.module.scss:47]: @grahamlangford With 1 less button on the main toolbar (see overflow changes) the justify space-between looked kind of weird. Taking this back out for now.

File: knip.mjs [hunk 7]
```diff
@@ -155,6 +155,7 @@ const knipConfig = {
         "@tiptap/react",
         "@tiptap/extension-underline",
         "@tiptap/extension-link",
+        "@tiptap/extension-image",
```
Comment (mnholtz) [inline knip.mjs:158]: I have no idea why dead-code is failing after adding this line

Comment (twschiller) [reply knip.mjs:158]: Wacky that it shows up under both unused dependencies and a configuration hint warning:

```
Unused dependencies (1)
@tiptap/extension-image  package.json
Configuration hints (6)
Unused item in ignoreDependencies (workspace: applications/browser-extension): @tiptap/extension-image
```

File: applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (grahamlangford) [reply applications/browser-extension/src/components/richTextEditor/toolbar/ImageButton/useUploadAsset.ts:81]: I would prefer all of this be written using `queryFn` rather than as a custom hook. Also, we should either place all calls in appAPI or all in the s3UploadAPI. I understand you have a separate api slice to use a different base query, but swapping out the base query is the first listed use for `queryFn` per the docs

File: applications/browser-extension/src/components/richTextEditor/toolbar/Toolbar.module.scss (hunk 6 above)
Comment (grahamlangford) [reply applications/browser-extension/src/components/richTextEditor/toolbar/Toolbar.module.scss:47]: Depending on look, the answer might just be to reduce the media query min-width, but I'll defer to you on what looks best. Only added it because `start` looked weird at very small widths

File: applications/browser-extension/src/components/IsolatedComponent.tsx [hunk 8]
```diff
@@ -110,6 +114,20 @@ type Props<T> = React.DetailedHTMLProps<
...
+    [appApi.reducerPath]: appApi.reducer,
+    [s3UploadApi.reducerPath]: s3UploadApi.reducer,
+  },
+  middleware(getDefaultMiddleware) {
+    /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
+    return getDefaultMiddleware()
+      .concat(appApi.middleware)
+      .concat(s3UploadApi.middleware);
+    /* eslint-enable unicorn/prefer-spread  */
+  },
```
Comment (grahamlangford) [inline applications/browser-extension/src/components/IsolatedComponent.tsx:128]: NIT: I don't think we want both, see my comment here: https://github.com/pixiebrix/pixiebrix-extension/pull/9513/files#r1842298998

Comment (grahamlangford) [inline applications/browser-extension/src/components/IsolatedComponent.tsx:128]: I'm not convinced we should have multiple api slices. See my other comment here: https://github.com/pixiebrix/pixiebrix-extension/pull/9513/files#r1842298998

We could also consider injecting the endpoints at runtime (https://redux-toolkit.js.org/rtk-query/api/created-api/code-splitting#injectendpoints), but that's probably overkill

File: applications/browser-extension/src/components/richTextEditor/RichTextEditor.test.tsx [hunk 9]
```diff
@@ -2,14 +2,31 @@ import React from "react";
...
 
 describe("RichTextEditor", () => {
   const user = userEvent.setup({
     delay: 20,
   });
 
+  const renderWithStore = (component: React.ReactElement) => {
+    const store = createTestStore();
+    return render(<Provider store={store}>{component}</Provider>);
+  };
```
Comment (grahamlangford) [inline applications/browser-extension/src/components/richTextEditor/RichTextEditor.test.tsx:26]: NIT: prefer the wrapper option for providers: https://testing-library.com/docs/react-testing-library/setup#custom-render

File: applications/browser-extension/src/components/richTextEditor/RichTextEditor.tsx [hunk 10]
```diff
@@ -20,23 +20,62 @@ import { EditorProvider, type EditorProviderProps } from "@tiptap/react";
...
+                      };
+                    },
+                  }).configure({
+                    assetDatabaseId,
+                    inline: true,
+                    HTMLAttributes: { style: "max-width: 100%" },
+                  }),
+                ]
+              : []),
+          ]}
```
Comment (grahamlangford) [inline applications/browser-extension/src/components/richTextEditor/RichTextEditor.tsx:69]: Let's extract this to a helper. I'm not in favor of this much inlined logic

File: applications/browser-extension/src/data/model/Asset.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (grahamlangford) [reply applications/browser-extension/src/data/model/Asset.ts:61]: @johnnymetz let's see if we can't get drf-spectacular prioritized once we're further along with the mono repo and the serializer swap

File: applications/browser-extension/src/sidebar/store.ts [hunk 11]
```diff
@@ -55,6 +55,7 @@ import { type SessionChangesState } from "@/store/sessionChanges/sessionChangesT
 import { type SettingsState } from "@/store/settings/settingsTypes";
 import { type SidebarState } from "@/types/sidebarTypes";
 import { persistSidebarConfig } from "@/store/sidebar/sidebarStorage";
+import { s3UploadApi } from "@/components/richTextEditor/toolbar/ImageButton/useUploadAsset";
```
Comment (grahamlangford) [inline applications/browser-extension/src/sidebar/store.ts:58]: Why is this needed for the sidebar store? I thought it was confined to the Rich Text Editor?

File: applications/browser-extension/src/components/IsolatedComponent.tsx [hunk 12]
```diff
@@ -151,8 +169,10 @@ export default function IsolatedComponent<T>({
     <EmotionShadowRoot mode={MODE} pb-name={name} {...props}>
       <style>{cssText}</style>
       <Stylesheets href={stylesheetUrl ?? []}>
-        {/* Must call the factory on each render to pick up changes to the component props */}
-        <Suspense fallback={null}>{factory(LazyComponent)}</Suspense>
+        <Provider store={store}>
+          {/* Must call the factory on each render to pick up changes to the component props */}
+          <Suspense fallback={null}>{factory(LazyComponent)}</Suspense>
+        </Provider>
```
Comment (grahamlangford) [inline applications/browser-extension/src/components/IsolatedComponent.tsx:175]: Does it make sense to have this at the IsolatedComponent level? Couldn't we push it down to the RichTextEditor?

Comment (grahamlangford) [reply applications/browser-extension/src/components/IsolatedComponent.tsx:175]: Scratch that, we don't want nested redux providers and that would happen in the Page Editor if we pushed it down

File: knip.mjs (hunk 7 above)
Comment (mnholtz) [reply knip.mjs:158]: > With npm workspace, workspace-specific dependencies need to be placed in the corresponding workspace

So when introducing new dependencies, is this a case where we cd into the browser-extension directory to `npm install @new-dependency`?

File: applications/browser-extension/src/components/richTextEditor/RichTextEditor.tsx (hunk 10 above)
Comment (mnholtz) [reply applications/browser-extension/src/components/richTextEditor/RichTextEditor.tsx]: Done, lmk what you think of the approach

File: applications/browser-extension/src/extensionConsole/pages/packageEditor/PackageHistory.test.tsx
Comment (grahamlangford) [inline applications/browser-extension/src/extensionConsole/pages/packageEditor/PackageHistory.test.tsx:1]: This and similar changes was required due to updating the swagger.ts

File: applications/browser-extension/src/components/richTextEditor/RichTextEditor.test.tsx [hunk 13]
```diff
@@ -188,4 +237,52 @@ describe("RichTextEditor", () => {
...
+
+    // Mock the asset update request
+    axiosMock.onPatch(API_PATHS.ASSET(databaseId, assetId)).reply(200, {
+      id: assetId,
+      download_url: mockDownloadUrl.href,
+      filename: file.name,
+      is_uploaded: true,
+      updated_at: new Date().toISOString(),
+      created_at: new Date().toISOString(),
+    });
```
Comment (grahamlangford) [inline applications/browser-extension/src/components/richTextEditor/RichTextEditor.test.tsx:275]: NIT: 
```suggestion
    axiosMock.onPost(API_PATHS.ASSET_PRE_UPLOAD(databaseId)).reply(200, {
      asset: {
        id: assetId,
        download_url: mockDownloadUrl.href,
        filename: file.name,
        is_uploaded: false,
        updated_at: new Date().toISOString(),
        created_at: new Date().toISOString(),
      },
      upload_url: mockUploadUrl.href,
      fields: {
        key: "test-key",
        policy: "test-policy",
      },
    }).onPost(mockUploadUrl.href)
    .reply(200)
    // Mock the asset update request
     .onPatch(API_PATHS.ASSET(databaseId, assetId)).reply(200, {
      id: assetId,
      download_url: mockDownloadUrl.href,
      filename: file.name,
      is_uploaded: true,
      updated_at: new Date().toISOString(),
      created_at: new Date().toISOString(),
    });
```

**Review by grahamlangford — APPROVED** [review]
